src/main/kotlin/org/rust/lang/core/psi/ext/RsBaseType.kt

```diff
@@ -6,14 +6,43 @@
 package org.rust.lang.core.psi.ext
 
 import org.rust.lang.core.psi.RsBaseType
+import org.rust.lang.core.psi.RsElementTypes.*
+import org.rust.lang.core.psi.RsPath
+import org.rust.lang.core.psi.tokenSetOf
 
-val RsBaseType.isCself: Boolean get() {
-    val path = path
-    return path != null && !path.hasColonColon && path.hasCself
+val RsBaseType.name: String? get() = path?.referenceName
+
+val RsBaseType.kind: RsBaseTypeKind
+    get() = when (kindEnum) {
+        RsBaseTypeKindEnum.UNIT -> RsBaseTypeKind.BtUnit
+        RsBaseTypeKindEnum.NEVER -> RsBaseTypeKind.BtNever
+        RsBaseTypeKindEnum.UNDERSCORE -> RsBaseTypeKind.BtUnderscore
+        RsBaseTypeKindEnum.PATH -> RsBaseTypeKind.BtPath(path ?: error("Malformed RsBaseType element: `$text`"))
+    }
+
+sealed class RsBaseTypeKind {
+    object BtUnit : RsBaseTypeKind()
+    object BtNever : RsBaseTypeKind()
+    object BtUnderscore : RsBaseTypeKind()
+    data class BtPath(val path: RsPath) : RsBaseTypeKind()
 }
 
-val RsBaseType.isUnit: Boolean get() = (stub?.isUnit) ?: (lparen != null && rparen != null)
-val RsBaseType.isNever: Boolean get() = (stub?.isNever) ?: (excl != null)
-val RsBaseType.isUnderscore: Boolean get() = (stub?.isUnderscore) ?: (underscore != null)
+private val RS_BASE_TYPE_KINDS = tokenSetOf(LPAREN, EXCL, UNDERSCORE, PATH)
 
-val RsBaseType.name: String? get() = path?.referenceName
+val RsBaseType.kindEnum: RsBaseTypeKindEnum get() {
+    val stub = stub
+    if (stub != null) return stub.kind
+
+    val child = node.findChildByType(RS_BASE_TYPE_KINDS)
+    return when (child?.elementType) {
+        LPAREN -> RsBaseTypeKindEnum.UNIT
+        EXCL -> RsBaseTypeKindEnum.NEVER
+        UNDERSCORE -> RsBaseTypeKindEnum.UNDERSCORE
+        PATH -> RsBaseTypeKindEnum.PATH
+        else -> error("Malformed RsBaseType element: `$text`")
+    }
+}
+
+enum class RsBaseTypeKindEnum {
```

Comment:
Awful name 😆 
Maybe use some other synonym of `kind`? Like `form` or something else (I can't come up with good synonym)? But I don't insist - anyway we usually shouldn't use anywhere except stubs

Reply:
Maybe RsBaseTypeStubKind?

Reply:
looks ok